Fix: serialization error when gossip working

kitsuyui opened the pull request:
Fix of https://github.com/celery/celery/issues/5075

## Current problem

Currently, Gossip does not respect accept_content when generating a consumer.
If you have multiple Celery workers running and allow pickle with accept_content, you may get a `Refusing to deserialize disabled content of type pickle (application/x-python-serialize)` error.

This error does not occur when running a single worker, but it does occur when running multiple workers.

## What this PR does

- Make Gossip respect accept_content
  - To do this, Gossip needs to pass accept_content when generating a consumer
- Add tests to verify that multiple workers are running and no errors containing `Refusing to deserialize disabled content of type` are occurring

## Additional notes

- Some of the existing unit tests use `Mock()` to replace Gossip processing. This is the case with `t/unit/worker/test_consumer.py`.
  - The `Mock()` needs to be configured with additional settings to work as intended with this change.

- Some of the tests are unstable. I (@kitsuyui) have confirmed that these failures occur without my changes.
1. The update to Codecov reports sometimes fails with a 500 error
2. Integration tests sometimes fail due to timeouts

auvipy commented on celery/worker/consumer/gossip.py:
```diff
@@ -176,6 +176,7 @@ def get_consumers(self, channel):
             channel,
             queues=[ev.queue],
             on_message=partial(self.on_message, ev.event_from_message),
+            accept=ev.accept,
```
hey, thanks for the PR, but this change needs some unit test to get accepted

auvipy commented:
> ## Description
> 
> fix of #5075
> Pass accept when Gossip getting consumers.

can you at least add a unit test for this fix?

kitsuyui commented:
@auvipy 
Thanks. I'd tried to do that but it will take more time for me.
Or can you write instead of me?

auvipy commented:
> @auvipy
> Thanks. I'd tried to do that but it will take more time for me.
> Or can you write instead of me?

you can take your time, I have other priorities todo in celery at the moment

thedrow commented:
@kitsuyui Is there a chance that you'll complete this PR soon?

kitsuyui commented:
@thedrow No.
I don't have any clue to solve soon. I've tried but the unit test has become too complicated.

jbvsmo commented:
Any news on this? As the author said he can't finish is it possible some celery mantainer pick it up, please? cc @auvipy 

This is affecting us greatly with no other fix or workaround

auvipy commented:
We need to verify this is working and not creating any regression. Thanks for the ping

rfqu commented:
I encountered this issue too and solved by adding 
`accept_content = ['application/json', 'application/x-python-serialize','application/x-pickle2']`

auvipy commented:
> I encountered this issue too and solved by adding `accept_content = ['application/json', 'application/x-python-serialize','application/x-pickle2']`

hey! great to know that. is it possible for you to send a PR to improve the docs regarding the issue/solution you mentioned?

wlach commented:
> I encountered this issue too and solved by adding `accept_content = ['application/json', 'application/x-python-serialize','application/x-pickle2']`

Based on what I've seen, I don't see how this could fix the problem since the problem is that gossip doesn't respect these settings. I'm pretty sure this PR is the correct solution based on a frustrating encounter with this issue today. 

Is there any chance we could merge this as-is and file a follow-up for a unit test? This is a pretty bad bug: it will affect anyone still using pickle serialization with gossip.

auvipy commented:
problem is the patch is making the CI red. which means it will create regression

kitsuyui commented:
Hi, sorry for being vague. Sorry for the long delay.

I have quitted the company that required this patch. However, I now have time to develop it. If there is anything I can do to help, I would like to do so.

I have already found a way to reproduce the problem at https://github.com/celery/celery/issues/5075#issuecomment-751647340, so I could try again.

The problem was that I was not able to write the necessary tests well, and I was unable to eliminate concerns about regression. I may still need someone's help in this.

auvipy commented:
thanks for coming back! as you shared in your repo https://github.com/kitsuyui/celery/pull/5#issuecomment-1774116373 some changes on these tests would be needed. I currently have a big backlog in celery to clear. so in the mean time if you can try to figure out some sort of test, it would be great

kitsuyui commented:
@auvipy Hi. I have updated the commits and description of the pull request.
I have added tests. I have also confirmed that the tests fail if I do not make this change.
I have confirmed that the tests pass on my local environment and on the CI of my fork.
But I am not confident that these tests meet your standards.
Could you please review this?

auvipy commented:
thanks for the tests! the failures seems to be github network related. so going to re start the CI

kitsuyui commented:
@auvipy Hi, I can't rerun CI because I don't have permissions on the project. Please retry. ☕

Nusnus commented:
@kitsuyui 
> @auvipy Hi, I can't rerun CI because I don't have permissions on the project. Please retry. ☕ 

Please rebase on `main`.
CI fixes were pushed.

I will make sure the CI will run if it won't automatically.

auvipy reviewed (changes requested):
please pull from main branch again

kitsuyui commented:
@Nusnus @auvipy Hi, I rebased the branch and pushed it again.

Nusnus commented:
> @Nusnus @auvipy Hi, I rebased the branch and pushed it again.

Thank you!
I'll review it later today

auvipy reviewed (approved):
looks good to me

kitsuyui commented:
@auvipy @Nusnus  Thank you!
